dbusmock/testcase.py

```diff
@@ -28,6 +30,260 @@
+        if timeout <= 0:
+            assert timeout > 0, f'timed out waiting for D-Bus object {path}: {last_exc}'
+
+
+class MockedDBus:
+    '''A D-Bus daemon instance that prepresents either a session or a system bus.
+
+    If used as a context manager it will automatically start the bus and clean up
+    after itself on exit:
+
+        >>> with MockedDBus(BusType.SESSION) as bus:
+        >>>    do_something(bus)
+
+    Otherwise, `start()` and `stop()` manually.
+    '''
+    def __init__(self, bustype: BusType):
+        self.bustype = bustype
+        self._daemon: Optional[subprocess.Popen] = None
+
+        self._datadir = Path(tempfile.mkdtemp(prefix='dbusmock_data_'))
+        self._socket = self._datadir / f"{self.bustype.value}_bus.socket"
+        subdir = "system-services" if bustype == BusType.SYSTEM else "services"
+        self._servicedir = self._datadir / subdir
+        self._servicedir.mkdir(parents=True)
+
+        self._config = self._servicedir / f'dbusmock_{self.bustype.value}_cfg'
+        self._config.write_text(f'''<!DOCTYPE busconfig PUBLIC "-//freedesktop//DTD D-Bus Bus Configuration 1.0//EN"
+     "http://www.freedesktop.org/standards/dbus/1.0/busconfig.dtd">
+    <busconfig>
+      <type>{self.bustype.value}</type>
+      <keep_umask/>
+      <listen>unix:path={self._socket}</listen>
+      <!-- We do not add standard_{self.bustype.value}_servicedirs (i.e. we only have our private services directory). -->
+      <servicedir>{self._servicedir}</servicedir>
+      <policy context="default">
+        <allow send_destination="*" eavesdrop="true"/>
+        <allow eavesdrop="true"/>
+        <allow own="*"/>
+      </policy>
+    </busconfig>
+    ''')
+
+    def __enter__(self) -> "MockedDBus":
+        # Allow for start() to be called manually even before the `with`
+        if self._daemon is None:
+            self.start()
+
+        return self
+
+    def __exit__(self,  exc_type, exc_val, exc_tb):
+        # Allow for stop() to be called manually within `with`
+        if self._daemon is not None:
+            self.stop()
+
+    @property
+    def address(self) -> str:
+        '''
+        Returns this D-Bus' address in the environment variable format, i.e. something like
+        unix:path=/path/to/socket
+        '''
+        assert self._daemon is not None, "Call start() first"
+        return f"unix:path={self._socket}" # type: ignore
+
+    @property
+    def servicedir(self) -> Path:
+        '''
+        The services directory (full path) for any ``.service`` files that need to be known to this
+        D-Bus.
+        '''
+        return self._servicedir
+
+    @property
+    def pid(self) -> int:
+        assert self._daemon is not None, "Call start() first"
+        return self._daemon.pid
+
+    def start(self):
+        '''Start the D-Bus daemon.'''
+        addr = f'unix:path={self._socket}'
+        argv = ['dbus-daemon', f'--address={addr}', f'--config-file={self._config}']
```

Comment:
actually, I'm pretty sure we don't need the `--address` arg here since we specify it in the given config file and `--address` is only needed to override that.

Reply:
removed now